### ISSUE TITLE
Enter tags without any delimiters in mix mode / Empty Pattern

### ISSUE THREAD

**Issue by KiranRW1** [yairEO/tagify#163]
Hi @yairEO ,

I loved your library and would like to use it for all my projects which would have this requirement of Tags, but on my tryouts I have encountered an issue which may totally be due to my lack of understanding, but to clarify does this library support empty pattern? 

I tried out a scenario with mode: 'mix' and added values for white-listing, and pattern to null etc. But if i don't set a valid pattern I am not getting the drop-down suggestions.

````
const tagify = new Tagify(document.querySelector('input[name=search-bar]'), {
            delimiters: null,
            tagTemplate: function (value, tagData) {
                return `
                <tag title='${value}'>
                    <x title=''></x>
                    <div>
                        <span class='tagify__tag-text'>${value}</span>
                    </div>
                </tag>`;
            },
            mode: 'mix',
            pattern    : /@/,
            whitelist: [
                { value: 'Afghanistan', code: 'AF' },
                { value: 'Åland Islands', code: 'AX' },
                { value: 'Albania', code: 'AL' },
                { value: 'Algeria', code: 'DZ' },
                { value: 'American Samoa', code: 'AS' },
                { value: 'Andorra', code: 'AD' },
                { value: 'Angola', code: 'AO' },
                { value: 'Anguilla', code: 'AI' },
                { value: 'Antarctica', code: 'AQ' },
                { value: 'Antigua and Barbuda', code: 'AG' },
                { value: 'Argentina', code: 'AR' },
                { value: 'Armenia', code: 'AM' },
                { value: 'Aruba', code: 'AW' },
                { value: 'Australia', code: 'AU' },
                { value: 'Austria', code: 'AT' },
                { value: 'Azerbaijan', code: 'AZ' }
            ],
            dropdown: {
                enabled: 1,
                classname: 'extra-properties',
                itemTemplate: function (tagData) {
                    return `<div class='tagify__dropdown__item ${tagData.class ? tagData.class : ""}'>
                        <span>${tagData.value}</span>
                    </div>`
                }
            },
            mapValueToProp: "code", // map tags' values to this property name, so this property will be the actual value and not the printed value on the screen
        });
````

**Comment by yairEO** [comment]
Hi, thank you for the nice words.

What are you trying to achieve? An auto-complete solution so when a word from the whitelist is starting to be typed, it will suggest an auto-completion from the dropdown? (without using any pattern)

**Comment by KiranRW1** [comment]
@yairEO Yup, exactly that. Any help would be very much appreciated.

**Comment by yairEO** [comment]
But do you want those words which are auto-completed to also looks visually like Tags?

**Comment by KiranRW1** [comment]
@yairEO So sorry that I couldn't reply back to you soon. Yes, of-course I want all properties of the tag, if a word in whitelist is chosen from dropdown.

**Comment by yairEO** [comment]
Ok, I will try to think of a solution for you

**Comment by yairEO** [comment]
If you would use this setting:

    dropdown: {
       enabled: 1,

Without any pattern, then on every keystroke you will see the suggestions dropdown (especially with [_fuzzy-search_ enabled](#164)) if there's a match with any whitelist items, and there will probably be matches so it would be an annoying user experience IMHO.

**Comment by privateOmega** [comment]
@yairEO Yes, I have that option set to 1 as well, but it just says you need at least 1 character to show the suggestions drop-down. But from what I understood from using your project is you need to set a starting pattern to enable suggestions drop-down. So that essentially means the inputs should be `@x // x is some character to show the dropdowns`. Please correct me if i am wrong.

This is the code I have currently
`
const tagify = new Tagify(document.querySelector('input[name=search-bar]'), {
            delimiters: null,
            tagTemplate: function (value, tagData) {
                return `
                <tag title='${value}'>
                    <x title=''></x>
                    <div>
                        <span class='tagify__tag-text'>${value}</span>
                    </div>
                </tag>`;
            },
            mode: 'mix',
            pattern: /@/,
            whitelist: [],
            dropdown: {
                enabled: 1,
                maxTags: 5,
                classname: 'extra-properties',
                itemTemplate: function (tagData) {
                    return `<div class='tagify__dropdown__item ${tagData.class ? tagData.class : ""}'>
                        <span>${tagData.value}</span>
                    </div>`
                }
            }
        })
`

**Comment by yairEO** [comment]
I know and fully understand what you were saying, but I think you didn't understand my last comment fully.. Imagine a person typing a dropdown appearing almost constantly with suggested words, changing its suggestions almost on every key typed. that might be annoying no?

**Comment by privateOmega** [comment]
@yairEO I missed that last line. Yes, of course I got your point. But wouldn't all those drop-down options be populated from white-list items? Then wouldn't it make sense to show those matched items. A great example I could think of is Autocomplete in Google search? Also talking about approximate matching, do we have any option to disable it, and use exact sub-string matching instead?

**Comment by yairEO** [comment]
and when should the logic show a suggestions dropdown?
This is quite a complex scenario..

For example, a user types:

> I like eating oranges but not tomatoes 

And the white list has `["oranges", "tomatoes"]`.

Does the code starts to look for a whitelist item if the user is typing and a space hasn't been typed yet? that might not work perfectly for whitelist items which has spaces in them.

For the above example, a user types:

> I

Then all the words which starts with `i` will be suggested

> I 

User hit spacebar, so now it *might* mean the user is done with the previous word/letter and will start a new one..

> I like

As the user typed the word `like` suggestions list is shown and narrows down on any items which contains `l`,`li`,`lik`,`like`.. 

------

The question is, which text to use for the search in the _whitelist_? should it be the last text typed after the last space (except the beginning word/character).

if we also want to include spaces in the search, this becomes tricky because what will be searched? any of words in the typed string that have any match with any of the words in the whitelist?

> I like eating 

should all the text _"i like eating"_ be matched against the whitelist? 

This feature request is quite complex

**Comment by yairEO** [comment]
For now I've setup a branch which breaks the words using spaces and lookup if there's any match with the whitelist:

https://github.com/yairEO/tagify/tree/feature/163

I've used this code for testing:

```
var tagify = new Tagify(document.querySelector('input[name=tags3-1]'), {
        mode       : 'mix',  // <--  Enable mixed-content
        pattern    : null,  // <--  Tag words which start with @ or # (can be a String instead of Regex)
        whitelist  : ['Homer', 'Marge', 'Bart', 'Lisa', 'Maggie', 'Mr. Burns', 'Ned', 'Milhouse', 'Moe'],
        dropdown   : {
            enabled : 1
        }
    })
```

Tell me what you think

**Comment by privateOmega** [comment]
@yairEO True, its a complex scenario. I will also do some thinking and get back to you with any thoughts. Thank you so much for the branch.

**Comment by privateOmega** [comment]
@yairEO So sorry that I couldn't get back to you any sooner with my thoughts, since I was on vacation. Like you said, I can't think of how we could tackle the situation with tag containing space. But I have tested out the branch you created for me, and it looks good for now. Maybe we could work to solve the aforementioned issue soon.
Could you help me out with something else?
The rendered tags look different from that in the examples.
![Uploading Capture.JPG…]()

**Comment by privateOmega** [comment]
@yairEO So sorry to bombard you with all of this, but I think there's a memory leak when you try to select a word separated by space from drop-down to build tag. I say that because on selecting space separated words for this branch, it causes my chrome instance to hang. I am assuming it might be because the script tries to fuzzy search for last parts from the selected tag.

**Comment by yairEO** [comment]
Hi, regarding the styling, I think it's a CSS issue most likely.. I cannot help without access to the page where Tagify was implemented. 

I also sadly have no time for this ticket since it is too big and i'm super busy with my day job at the moment.. maybe in a month or two I'll have more time..
Your best option is to debug that branch and see what's making Chrome hang. 

I would also suggest merging Tagify's `master` branch with that experimental branch to keep fixes in sync.

**Comment by privateOmega** [comment]
@yairEO Looks like it, but even I am not sure what seems to be causing the problem. I understand, thanks. Sure will merge the master to that branch.

**Comment by jnpatrick99** [comment]
I tried the new code as well and it does indeed freeze the Chrome. Process starts to eat 99% CPU after clicking on any tag in the dropdown.

**Comment by privateOmega** [comment]
@jnpatrick99 But for me, it just occurs rarely. I just kill the tab and open the page in new tab and mostly it doesn't occur again. And also there seems to be some dom exception happening, but because of my less expertise, I am unable to solve it. Would be great if you could take a look at the code and solve it.

**Comment by jnpatrick99** [comment]
@privateOmega I think an endless loop starts as long as you supply null to the pattern and apply the feature 163 (https://github.com/yairEO/tagify/tree/feature/163) to the code in the mix mode.

Only @yairEO can help us to resolve it! I didn't have too much success finding the cause.

**Comment by yairEO** [comment]
@jnpatrick99 

I have no idea how can the `while` loop possibly cause and endless loop.. I cannot reproduce the bug.

Anyway, I've [implemented a "fix"](https://github.com/yairEO/tagify/blob/master/src/tagify.js#L821) in case it does occur.

Regarding your second update, i do not see anywhere in the code `tagValue = tag[0]` nor `tagValue`

**Comment by privateOmega** [comment]
@yairEO He's talking with respect to `feature/163 branch` which you created for me. In that branch, you have created a variable `tagValue `and when there's no pattern, you assign `tagValue = tag[0]`

**Comment by privateOmega** [comment]
> Hi, regarding the styling, I think it's a CSS issue most likely.. I cannot help without access to the page where Tagify was implemented.
> 
> I also sadly have no time for this ticket since it is too big and i'm super busy with my day job at the moment.. maybe in a month or two I'll have more time..
> Your best option is to debug that branch and see what's making Chrome hang.
> 
> I would also suggest merging Tagify's `master` branch with that experimental branch to keep fixes in sync.

I found out that the [Bulma](https://github.com/jgthms/bulma) CSS framework I was using had overridden the some CSS property for tag. I fixed that issue by adding `line-height    : 1.3;` under tag class.

**Comment by yairEO** [comment]
Already done apparently. use `dropdown:{enabled:0}` in your Tagify *settings* 

closing
